storage/remote/read_test.go

```diff
@@ -231,182 +211,310 @@ func TestSeriesSetFilter(t *testing.T) {
+
+		expectedQuery  *prompb.Query
+		expectedSeries []labels.Labels
 	}{
 		{
-			requiredMatchers: []*labels.Matcher{},
+			name: "empty",
+			mint: 1, maxt: 2,
 			matchers: []*labels.Matcher{
-				labels.MustNewMatcher(labels.MatchEqual, "special", "label"),
+				labels.MustNewMatcher(labels.MatchNotEqual, "a", "something"),
+			},
+			readRecent: true,
+
+			expectedQuery: &prompb.Query{
+				StartTimestampMs: 1,
+				EndTimestampMs:   2,
+				Matchers: []*prompb.LabelMatcher{
+					{Type: prompb.LabelMatcher_NEQ, Name: "a", Value: "something"},
+				},
+			},
+			expectedSeries: []labels.Labels{
+				labels.FromStrings("a", "b"),
+				labels.FromStrings("a", "b2", "region", "europe"),
+				labels.FromStrings("a", "b3", "region", "us"),
 			},
-			seriesSet: mockSeriesSet{},
 		},
 		{
-			requiredMatchers: []*labels.Matcher{
-				labels.MustNewMatcher(labels.MatchEqual, "special", "label"),
-			},
+			name: "external labels specified, not explicitly requested",
+			mint: 1, maxt: 2,
 			matchers: []*labels.Matcher{
-				labels.MustNewMatcher(labels.MatchEqual, "special", "label"),
+				labels.MustNewMatcher(labels.MatchNotEqual, "a", "something"),
+			},
+			readRecent: true,
+			externalLabels: labels.Labels{
+				{Name: "region", Value: "europe"},
+			},
+
+			expectedQuery: &prompb.Query{
+				StartTimestampMs: 1,
+				EndTimestampMs:   2,
+				Matchers: []*prompb.LabelMatcher{
+					{Type: prompb.LabelMatcher_NEQ, Name: "a", Value: "something"},
+					{Type: prompb.LabelMatcher_EQ, Name: "region", Value: "europe"},
+				},
+			},
+			expectedSeries: []labels.Labels{
+				labels.FromStrings("a", "b"),
+				labels.FromStrings("a", "b2"),
 			},
-			seriesSet: mockSeriesSet{},
 		},
 		{
-			requiredMatchers: []*labels.Matcher{
-				labels.MustNewMatcher(labels.MatchEqual, "special", "label"),
+			name: "external labels specified, explicitly requested europe",
+			mint: 1, maxt: 2,
+			matchers: []*labels.Matcher{
+				labels.MustNewMatcher(labels.MatchNotEqual, "a", "something"),
+				labels.MustNewMatcher(labels.MatchEqual, "region", "europe"),
+			},
+			readRecent: true,
+			externalLabels: labels.Labels{
+				{Name: "region", Value: "europe"},
+			},
+
+			expectedQuery: &prompb.Query{
+				StartTimestampMs: 1,
+				EndTimestampMs:   2,
+				Matchers: []*prompb.LabelMatcher{
+					{Type: prompb.LabelMatcher_NEQ, Name: "a", Value: "something"},
+					{Type: prompb.LabelMatcher_EQ, Name: "region", Value: "europe"},
+				},
+			},
+			expectedSeries: []labels.Labels{
+				labels.FromStrings("a", "b"),
+				labels.FromStrings("a", "b2", "region", "europe"),
 			},
+		},
+		{
+			name: "external labels specified, explicitly requested not europe",
+			mint: 1, maxt: 2,
 			matchers: []*labels.Matcher{
-				labels.MustNewMatcher(labels.MatchRegexp, "special", "label"),
+				labels.MustNewMatcher(labels.MatchNotEqual, "a", "something"),
+				// TODO(bwplotka): Is this what we want? I prometheus is region=europe and user requests region=us we
```

Comment:
A bit surprised with this. Is this intended (this PR does not change any semantic, just I added a test case for this), @brian-brazil ? :thinking: